File: POO/Tabuleiro_python/Tabuleiro.py
import math

class Tabuleiro:

    # Constructor
    def __init__(self, valores):
        # Calcula o tamanho do tabuleiro (lado)
        self.tam_tab = int(math.sqrt(len(valores)))

        # Cria a matriz que representa o tabuleiro
        self.tab = [valores[i*self.tam_tab : (i+1)*self.tam_tab] for i in range (self.tam_tab)]
        
        # Encontra a posição do zero
        self.zero_l, self.zero_c = self.encontra_zero((self.tab))

    # Função para verificar se o tabuleiro está na ordem correta
    def verifica_tab(self):
        for i, linha in enumerate(self.tab):
            for j, valor in  enumerate(linha):
                # Condição de alguma peça estar fora do lugar
                if valor != self.tam_tab*i + j:
                    return False
        return True
    
    def encontra_zero(self, tab):
        for i, linha in enumerate(tab):
            for j, valor in enumerate(linha):
                if valor == 0:
                    return i, j

    def move(self, sequencia):
        for i in range (len(sequencia)):
            self.aplica_movimento(sequencia[i])
            self.imprime_tab()

    def aplica_movimento(self, mov):
        # up
        if mov == 'u':
            if self.zero_l < (self.tam_tab - 1):
                self.tab[self.zero_l][self.zero_c] = self.tab[self.zero_l+1][self.zero_c]
                self.tab[self.zero_l+1][self.zero_c] = 0
                self.zero_l = self.zero_l + 1
        # down
        elif mov == 'd':
            # Trocar o 0 com o de cima
            if self.zero_l > 0 :
                self.tab[self.zero_l][self.zero_c] = self.tab[self.zero_l-1][self.zero_c]
                self.tab[self.zero_l-1][self.zero_c] = 0
                self.zero_l = self.zero_l - 1
         
        # left
        elif mov == 'l':
            # Trocar o 0 com o da direita
            if self.zero_c < self.tam_tab-1 :
                self.tab[self.zero_l][self.zero_c] = self.tab[self.zero_l][self.zero_c+1]
                self.tab[self.zero_l][self.zero_c+1] = 0
                self.zero_c = self.zero_c + 1
         
        # right
        elif mov == 'r':
            # Trocar o 0 com o da esquerda
            if self.zero_c > 0 :
                self.tab[self.zero_l][self.zero_c] = self.tab[self.zero_l][self.zero_c-1]
                self.tab[self.zero_l][self.zero_c-1] = 0
                self.zero_c = self.zero_c - 1
         
    # Função que imprime na tela o estado atual do tabuleiro
    def imprime_tab(self):
        for i in range (self.tam_tab): 
            for _ in range (self.tam_tab): 
                print("+------", end="")
            print("+")
            print("|  ", end="")
            
            for j in range (self.tam_tab):
                if self.tab[i][j] < 10 :
                    print(" ", end="")   # Consertar a formatação
                if self.tab[i][j] != 0 :
                    print(self.tab[i][j], end = "")
                else:
                    print(" ", end="")
                print("  |  ", end="")
                   
            print("\n", end="")
        
        for _ in range (self.tam_tab):
            print("+------", end="")
        print("+\n")
        

# [0, 1, 2, 3, 4, 5, 6, 7, 8]

File: POO/Tabuleiro_java/Tabuleiro.java
public class Tabuleiro {
    private int[][] tab;
    private int tam_tab;
    private int zero_l;
    private int zero_c;

    // Construtor
    public Tabuleiro(int[] valores){
        // Calcula o tamanho do tabuleiro
        tam_tab = (int)Math.sqrt(valores.length); 
        tab = new int[tam_tab][tam_tab];

        // Passa os valores para a matriz que representa o tabuleiro
        int k = 0;
        for(int i=0; i<tam_tab; i++){
            for(int j=0; j<tam_tab; j++){
                tab[i][j] = valores[k++];
            }
        }

        // Guarda a posição do zero no tabuleiro
        for(int i=0; i<tam_tab; i++){
            for(int j=0; j<tam_tab; j++){
                if (tab[i][j] == 0){
                    zero_l = i;
                    zero_c = j;
                    return;
                }
            }
        }
    }   

    // Função que verifica se o tabuleiro está na ordem correta
    public Boolean verifica_tab(){
        for(int i=0; i<tam_tab; i++){
            for(int j=0; j<tam_tab; j++){
                // Condição de alguma peça no lugar errado

                if (tab[i][j] != tam_tab*i+j) return false;
            }
        }
        return true;
    }

    // Função que decodifica a string recebida e chama a função que realiza o movimento
    public void move(String seq){
        for(int i=0; i<seq.length(); i++){
            char mov = seq.charAt(i);
            aplica_movimento(mov);
            imprime_tab();
        }
    }

    // Função que aplica o movimento no tabuleiro de acordo com o caractere recebido
    public void aplica_movimento(char mov){
        // up
        if(mov == 'u'){
            // Trocar o 0 com o de baixo
            if(zero_l < tam_tab-1){
                tab[zero_l][zero_c] = tab[zero_l+1][zero_c];
                tab[zero_l+1][zero_c] = 0;
                zero_l++; 
            }
        }
        // down
        else if(mov == 'd'){
            // Trocar o 0 com o de cima
            if(zero_l > 0){
                tab[zero_l][zero_c] = tab[zero_l-1][zero_c];
                tab[zero_l-1][zero_c] = 0;
                zero_l--;
            }
        }
        // left
        else if(mov == 'l'){
            // Trocar o 0 com o da direita
            if(zero_c < tam_tab-1){
                tab[zero_l][zero_c] = tab[zero_l][zero_c+1];
                tab[zero_l][zero_c+1] = 0;
                zero_c++;
            }
        }
        // right
        else if(mov == 'r'){
            // Trocar o 0 com o da esquerda
            if(zero_c > 0){
                tab[zero_l][zero_c] = tab[zero_l][zero_c-1];
                tab[zero_l][zero_c-1] = 0;
                zero_c--;
            }
        }
    }

    // Função que imprime na tela o estado atual do tabuleiro
    public void imprime_tab(){
        for(int i=0; i<tam_tab; i++){
            for(int k=0; k<tam_tab; k++) System.out.print("+------");
            System.out.println("+");
            System.out.print("|  ");
            
            for(int j=0; j<tam_tab; j++){
                if(tab[i][j] < 10)System.out.print(" ");   // Consertar a formatação
                System.out.print(tab[i][j]);
                System.out.print("  |  ");
                
            }   
            System.out.print("\n");
        }
        for(int k=0; k<tam_tab; k++) System.out.print("+------");
        System.out.println("+\n");
    }

}
